implementation branch review

locorocorolling opened the pull request.

Ricky0625 committed:
feat: loading state

Zyeoh883 committed:
Login routes the user to the approriate pages. 2FA Page sends an OTP to the user's email upon entering the webpage, a resend button with a cooldown of 5 seconds is also made. There's some funny interaction with the cookies where there are force routes on certain webpages that will be fixed later.

ziqiqiiii committed:
[DONE] upload images to Azure Blob Storage

ziqiqiiii committed:
[UPDATE] installed dependent packages

Zyeoh883 committed:
bug: added middleware but now 2FA does not enter admin_dashboard

OTP_verification is supposed to be called in 2FA page and it also deletes previous OTPs if successful. Unit testing using Postman shows OTP_verification works but implementation is not working as the API is never called or never successful? Once a valid OTP is given on the 2FA page, user is routed to login page for some reason.

ziqiqiiii committed:
[UPDATE] Fixed the typescript problem

ziqiqiiii committed:
Enhanced the code for sasToken and uploadImage

ziqiqiiii committed:
Deleted unnecessary code

ziqiqiiii committed:
[REFACTOR] Move the utils directory under src directory

ziqiqiiii committed:
[UPDATE && REFACTOR] Added UploadImageRequest Interface and refactor the response code

ziqiqiiii committed:
[DELETE] Deleted unnecessary file

Ricky0625 committed:
fix: updated admin_2FA

- add onSubmit event on form, and trigger handleSubmit
- call the correct api: createSessionID -> SessionID
- call setEmail

Ricky0625 committed:
fix: update SessionID route

Disable secure:true. This should be enabled in production only.

Ricky0625 committed:
chore: removed console.log

Ricky0625 committed:
chore: new package: cookies

Zyeoh883 committed:
fix: Middleware is now working to redirect away from admin_dashboard

Had to add the following code into the middleware call for the SessionID GET
request:
'Cookie': `sessionID=${sessionID}`

Also removed the messages for check if sessionExists in SessionID GET to
get it working without prompting the messages. Will add back later.

Zyeoh883 committed:
feat: middleware now handles routes better

Not fully completed but it now reroutes a valid admin from login and 2FA
to dashabord.
API calls (only /admin/Dashboard) are protected and routes
user to login page.

Zyeoh883 committed:
feat: Logout now clears SessionID from cookies and reroutes to '/'

Added DELETE api for cookies to delete locally and on MongoDB.
Edited DropdownMenu.tsx to call DELETE api in /api/admin/SessionID.

Zyeoh883 committed:
bugs: 'useReducer' Error

Archiving this commit to check previous commits.
Attempting to simplify middleware by calling a validateSession function
rather than api call. Added the function as /lib/validateSession.ts.

Ricky0625 committed:
Merge pull request #13 from 42KL-admin/SasTokenToUploadImages

sasToken to upload images to Azure Blob Storage

Zyeoh883 committed:
feat: improvements in middleware redirects

Abstracted middleware admin pages logic to /lib/middleware/validateSession.ts.
Remove routing in src/app/(authenticated)/layout.tsx.
Added router.push to '/' in admin_dashboard page and @hero.

Ricky0625 committed:
Merge pull request #14 from 42KL-admin/frontend/result

feat: loading state

Ricky0625 committed:
chore: add bun.lockb to gitignore

Ricky0625 committed:
feat: apiCall wrapper function

Ricky0625 committed:
chore: refactored isSasTokenExpired as util function

Ricky0625 committed:
feat: change api/sasToken api

- api will not return a sasToken anymore
- api will set the sasToken as a cookie in the server side
- api will now responsible to check whether the token is expired or not
- api will regenerate token if expired
- api will use the existing token if the previous token is not expired

Ricky0625 committed:
feat: add services folder

- services folder will consists of the wrapper function of server API
- added wrapper function to ask server to check Sas token

Ricky0625 committed:
fix(api/uploadImage): read sasToken from cookies

Zyeoh883 committed:
fix: Fixed where EmailOtp is not sent

Middleware prevented access to EmailOtp, need to check the api call if
it was called at Admin_2FA

Ricky0625 committed:
chore: make apiCall function more generic

Ricky0625 committed:
fix: remove sasToken from UploadFilesRequest interface

Ricky0625 committed:
chore: remove un-used console log

Ricky0625 committed:
feat: uploadSlateApi wrapper function

Ricky0625 committed:
fix: remove the need to pass in sasToken for post request, read from cookie instead

Ricky0625 committed:
feat: upload image to blob storage

Ricky0625 committed:
feat: rename file to timestamp_file to avoid file overwrite

Zyeoh883 committed:
fix: Fixed run build errors

timeoutID in 2FA_page is adjusted to prevent type errors.
Added router in the dependency array in 2FA layout.
Changed back tick in Admin_dashboard

Ricky0625 committed:
Merge pull request #15 from 42KL-admin/frontend/sas_token

feat: Implement SAS token and Upload Image API

locorocorolling commented on src/app/api/admin/SessionID/route.ts:
```diff
@@ -0,0 +1,100 @@
+			return NextResponse.json({ message: "SessionID not found" }, { status: 404 });
+		}
+
+		const response = NextResponse.json({ message: "SessionID deleted successfully" }, { status: 200 });
```
@Zyeoh883  
for Line 58,  
i just realised it's actually two separate arguments (note the comma in between!)
![brave_ui3HiOAQTo](https://github.com/42KL-admin/42DoGood-Reef_Check/assets/20178082/de435caf-3020-4b6a-a785-53b36dc72f92)

hovering over the status property/key also shows it's part of `ResponseInit`, that's why we couldn't read `result.status` in the `DropdownMenu.tsx` file
![image](https://github.com/42KL-admin/42DoGood-Reef_Check/assets/20178082/edb86f80-bbf8-4fa9-a8a9-36daf4932d35)
_(line 58)_



`response.status` does work as expected
![image](https://github.com/42KL-admin/42DoGood-Reef_Check/assets/20178082/ee5ab540-a021-4e56-a5c2-06820db077eb)
 _(DropdownMenu.tsx)_
![image](https://github.com/42KL-admin/42DoGood-Reef_Check/assets/20178082/0da9c2ae-3e57-4f71-a45f-abf1762eafe5)

this should solve the problem!

locorocorolling commented:
it's possible to manually change the cookies' `sessionID` to something invalid, and still access the admin pages,  
i blame Next.js for having its own version of `NextRequest` and `NextResponse`
https://github.com/42KL-admin/42DoGood-Reef_Check/assets/20178082/1a526a6e-1856-4d82-815e-9a4cffe9581d

similar issue as above, misunderstanding the NextResponse's arguments / API:
![image](https://github.com/42KL-admin/42DoGood-Reef_Check/assets/20178082/dbd30c08-3cba-4ddd-9e53-ddfd41dcd71f)
_(sessionID/route.ts)_


in this case, it's returning a response with `200 OK`, that has a **json body of `{ "status": 401 }`**
![Cursor_B10FZ7fdLP](https://github.com/42KL-admin/42DoGood-Reef_Check/assets/20178082/0a5811c1-b205-4cec-bce3-4ae43c001029)
_(middleware.ts)_

![image](https://github.com/42KL-admin/42DoGood-Reef_Check/assets/20178082/7c2311d3-760b-497f-a62b-ea9445f651cb)


so yeah! this should help too!

Zyeoh883 reviewed (approved):
Good spot on issues!!! That should fix most of the problems, I'll append some changes to the middleware after the merge.

Zyeoh883 committed:
fix: DropdownMenu now logouts correctly

logoutUser gets the status from the response directly.

Zyeoh883 committed:
Merge branch 'dev' into implementation

Zyeoh883 committed:
chore: updated package-lock.json

Zyeoh883 committed:
fix: fixed dependencies for build

Zyeoh883 committed:
fix: EmailOtp & OtpVerification is fixed

Both EmailOtp and OTPVerification api is now working, bcrypt.compare
should not be used as == is sufficient.

Attempting to reroute admin to home page if sessionID is invalid when
using the components in the dashbaord page. It console.logs that it is
in home but it remains in Daashboard until the refresh?
At least it prevents invalid SessionID from touching anything

Invalid SessionID will prompt the middleware to delete all cookies and
attempt to route to home.

Zyeoh883 committed:
chore: Switched from 2 tabs to 2 spaces

.prettierrc.json is updated to have 2 spaces rather than 2 tabs for
width.

Zyeoh883 committed:
chore: updated package-lock.json

Zyeoh883 committed:
fix: Fix DropdownMenu Component not loging out for users

The logoutUser calls a DELETE api to delete SessionID even for users. It
fails then an error is thrown before it can clear and redirect.
Made it where it deletes the email cookie first, calls the DELETE api if
admin and then redirects to home.
